addons/bus/static/src/services/worker_service.js

```diff
@@ -0,0 +1,132 @@
+import { timings } from "@bus/misc";
+import { browser } from "@web/core/browser/browser";
+import { registry } from "@web/core/registry";
+import { Deferred } from "@web/core/utils/concurrency";
+import { session } from "@web/session";
+
+export const WORKER_STATE = Object.freeze({
+    UNINITIALIZED: "UNINITIALIZED",
+    INITIALIZING: "INITIALIZING",
+    INITIALIZED: "INITIALIZED",
+    FAILED: "FAILED",
+});
+
+export class WorkerService {
+    constructor(env, services) {
+        this.params = services["bus.parameters"];
+        this.worker = null;
+        this.isUsingSharedWorker = Boolean(browser.SharedWorker);
+        this._state = WORKER_STATE.UNINITIALIZED;
+        this.connectionInitializedDeferred = null;
+        this.throttledStartWorker = timings.throttle(() => this.startWorker(), 1000);
+    }
+
+    startWorker() {
+        if (!this.connectionInitializedDeferred) {
+            this.connectionInitializedDeferred = new Deferred();
+        }
```

Comment:
I think that with the changes we've made, we can just initialize the `connectionInitializedDeferred` property to `new Deferred()` right away.

Reply:
we actually need this. When a shared worker failed, we will then call startworker again, in this case we shouldn't create a new deferred